drivers/tee/optee/optee.c

```diff
@@ -150,6 +150,24 @@ static int param_to_msg_param(const struct tee_param *param, unsigned int num_pa
 	return 0;
 }
 
+static void msg_param_to_tmp_mem(struct tee_param *p, uint32_t attr,
+				 const struct optee_msg_param *mp)
+{
+	struct tee_shm *shm = (struct tee_shm *)mp->u.tmem.shm_ref;
+
+	p->attr = TEE_PARAM_ATTR_TYPE_MEMREF_INPUT + attr - OPTEE_MSG_ATTR_TYPE_TMEM_INPUT;
+	p->b = mp->u.tmem.size;
+
+	if (!shm) {
+		p->a = 0;
+		p->c = 0;
+		return;
+	}
+
+	p->a = mp->u.tmem.buf_ptr - z_mem_phys_addr(shm->addr);
```

Comment:
It would be nice to add some comments about `a`, `b`, `c` parameters use in this case. Because I honestly does not understand what is happening here.

You are subtracting physical address from a virtual address?

Reply:
It appears to be already documented in zephyr/include/drivers/tee.h:

>  225 /**                                                                                                                                                                                                                                                                                                                                                                   
>  226  * struct tee_param - parameter                                                                                                                                                                                                                                                                                                                                       
>  227  * @attr: attributes                                                                                                                                                                                                                                                                                                                                                  
>  228  * @a: if a memref, offset into the shared memory object, else a value parameter                                                                                                                                                                                                                                                                                      
>  229  * @b: if a memref, size of the buffer, else a value parameter                                                                                                                                                                                                                                                                                                        
>  230  * @c: if a memref, shared memory identifier, else a value parameter                                                                                                                                                                                                                                                                                                  
>  231  *                                                                                                                                                                                                                                                                                                                                                                    
>  232  * @attr & TEE_PARAM_ATTR_TYPE_MASK indicates if memref or value is used in                                                                                                                                                                                                                                                                                           
>  233  * the union. TEE_PARAM_ATTR_TYPE_VALUE_* indicates value and                                                                                                                                                                                                                                                                                                         
>  234  * TEE_PARAM_ATTR_TYPE_MEMREF_* indicates memref. TEE_PARAM_ATTR_TYPE_NONE                                                                                                                                                                                                                                                                                            
>  235  * indicates that none of the members are used.                                                                                                                                                                                                                                                                                                                       
>  236  *                                                                                                                                                                                                                                                                                                                                                                    
>  237  * Shared memory is allocated with TEE_IOC_SHM_ALLOC which returns an                                                                                                                                                                                                                                                                                                 
>  238  * identifier representing the shared memory object. A memref can reference                                                                                                                                                                                                                                                                                           
>  239  * a part of a shared memory by specifying an offset (@a) and size (@b) of                                                                                                                                                                                                                                                                                            
>  240  * the object. To supply the entire shared memory object set the offset                                                                                                                                                                                                                                                                                               
>  241  * (@a) to 0 and size (@b) to the previously returned size of the object.                                                                                                                                                                                                                                                                                             
>  242  */
This tee_param may be considered as a structure for client -> driver communication purposes. Unfortunately I couldn't find any mentions in the TEE documentation to regulate those parameters. That's why I consider it implementation specific. During research I've found that Linux Kernel uses similar implementation. But it uses union in tee_param structure, which divides value parameter and memref object.

Reply:
Hm...  it's already documented in tee.h header in the description of the structure. Isn't that enough?

Reply:
Sorry, missed that. My bad.